src/slicer.cpp

```diff
@@ -787,72 +787,59 @@ void SlicerLayer::makePolygons(const Mesh* mesh)
 
     TimeKeeper slice_timer;
 
-    layers.resize(slice_layer_count);
+    layers = 
+        buildLayersWithHeight(slice_layer_count, slicing_tolerance, initial_layer_thickness, thickness, 
+            use_variable_layer_heights, adaptive_layers);
 
-    // set (and initialize compensation for) initial layer, depending on slicing mode
-    const coord_t initial_layer_thickness = Application::getInstance().current_slice->scene.current_mesh_group->settings.get<coord_t>("layer_height_0");
-    layers[0].z = std::max(0LL, initial_layer_thickness - thickness);
-    coord_t adjusted_layer_offset = initial_layer_thickness;
-    if (use_variable_layer_heights)
-    {
-        layers[0].z = adaptive_layers->at(0).z_position;
-    }
-    else if (slicing_tolerance == SlicingTolerance::MIDDLE)
-    {
-        layers[0].z = initial_layer_thickness / 2;
-        adjusted_layer_offset = initial_layer_thickness + (thickness / 2);
-    }
 
-    // define all layer z positions (depending on slicing mode, see above)
-    for (unsigned int layer_nr = 1; layer_nr < slice_layer_count; layer_nr++)
-    {
-        if (use_variable_layer_heights)
-        {
-            layers[layer_nr].z = adaptive_layers->at(layer_nr).z_position;
-        }
-        else
-        {
-            layers[layer_nr].z = adjusted_layer_offset + (thickness * (layer_nr - 1));
-        }
-    }
+    std::vector<std::pair<int32_t, int32_t>> zbbox = buildZHeightsForFaces(*mesh);
 
-    // loop over all mesh faces
-    for (unsigned int mesh_idx = 0; mesh_idx < mesh->faces.size(); mesh_idx++)
-    {
-        // get all vertices per face
-        const MeshFace& face = mesh->faces[mesh_idx];
-        const MeshVertex& v0 = mesh->vertices[face.vertex_index[0]];
-        const MeshVertex& v1 = mesh->vertices[face.vertex_index[1]];
-        const MeshVertex& v2 = mesh->vertices[face.vertex_index[2]];
 
-        // get all vertices represented as 3D point
-        Point3 p0 = v0.p;
-        Point3 p1 = v1.p;
-        Point3 p2 = v2.p;
+    buildSegments(*mesh, zbbox, layers);
 
-        // find the minimum and maximum z point
-        int32_t minZ = p0.z;
-        int32_t maxZ = p0.z;
-        if (p1.z < minZ) minZ = p1.z;
-        if (p2.z < minZ) minZ = p2.z;
-        if (p1.z > maxZ) maxZ = p1.z;
-        if (p2.z > maxZ) maxZ = p2.z;
+    log("slice of mesh took %.3f seconds\n", slice_timer.restart());
+
+    makePolygons(*mesh, slicing_tolerance, layers);
+    log("slice make polygons took %.3f seconds\n", slice_timer.restart());
+}
 
-        // calculate all intersections between a layer plane and a triangle
-        for (unsigned int layer_nr = 0; layer_nr < layers.size(); layer_nr++)
+void Slicer::buildSegments(const Mesh &mesh, const std::vector<std::pair<int32_t, int32_t>> &zbbox,
+    std::vector<SlicerLayer>& layers) 
+{
+    // OpenMP
+#pragma omp parallel for default(none) shared(mesh, zbbox, layers)
+    for (int layer_nr = 0; layer_nr < layers.size(); layer_nr++)
```

Comment:
Comparing integers of different signedness (`Wsign-compare` in GCC).

Reply:
I think the best option here is a `LayerIndex` because it's very semantic, but then you'd need a cast to tell the compiler that you're working with layer indices here because that vector doesn't return `LayerInd`ices.

Reply:
I am afraid, OpenMP is not compatible with LayerIndex (at lease for MSVC).
BTW, what do you think about the new standard "std::for_each(std::execution::par,...)"?"

Reply:
We compile CuraEngine with GCC though, not MSVC. Nice to hear though that MSVC is still working (for the rest then). Alternatively you can use int and cast the size to an int, of course. I can't guarantee that we won't give a for loop a LayerIndex in the future though, since we don't test (or support) MSVC.

I think that `std::for_each` with parallel execution is a very nice alternative, if it works for us. It'd also work with MacOS, which is as of now still single-threaded. I hope it works fine together with OpenMP so that it can be transitioned softly rather than having to make a hard atomic transition.
Currently we're still compiling with C++11 though. Hopefully soon that will change to C++17.

Reply:
Apart from GCodeLayerThreader, CuraEngine parallization is just about loops and 
these loops are executed "sequentially". So the migration ought to be quite smooth. Maybe even GCodeLayerThreader could be converted to a simple loop...
Is there a specific time line to upgrade to C++17?

Reply:
Not really since we don't generally work with timelines and promises to fix something before a certain deadline. It was recently brought up again and moved up a bit in our backlog. Most likely we'd plan this in after the 4.5 branch has split off to reduce risk for that release (though personally I won't think there's any risk at all).

Reply:
Thanks. I am looking forward!